File: elixir-impl/lib/task_sorter.ex
defmodule JobProcessor.TaskSorter do
  @moduledoc """
  Sorts tasks based on their dependencies.
  """

  def sort(tasks) when is_list(tasks) do
    with graph <- build_tasks_graph(tasks),
         sorted_task_names when is_list(sorted_task_names) <- Graph.topsort(graph) do
      map_sorted_names_to_tasks(sorted_task_names, tasks)
    else
      _ -> false
    end
  end

  defp build_tasks_graph(tasks) do
    tasks
    |> Enum.reduce(Graph.new(), fn task, graph ->
      graph = Graph.add_vertex(graph, task["name"], task)
      requires = Map.get(task, "requires", [])
      add_edges(graph, task["name"], requires)
    end)
  end

  defp add_edges(graph, task_name, requires) do
    Enum.reduce(requires, graph, fn dep, graph ->
      Graph.add_edge(graph, dep, task_name)
    end)
  end

  defp map_sorted_names_to_tasks(sorted_task_names, tasks) do
    task_map = Map.new(tasks, &{&1["name"], &1})
    Enum.map(sorted_task_names, &Map.get(task_map, &1))
  end
end

File: erlang-impl/src/task_sorter.erl
-module(task_sorter).
-export([sort/1]).

%% @doc Sorts the given list of tasks based on their dependencies.
-spec sort(Tasks :: [map()]) -> {ok, [map()]} | {error, binary()}.
sort(Tasks) ->
    G = digraph:new(),
    try
        add_tasks(Tasks, G),
        case digraph_utils:topsort(G) of
            false ->
                {error, error_message:circular_dependency()};
            SortedNames ->
                TaskMap = create_task_map(Tasks),
                {ok, [maps:get(Name, TaskMap) || Name <- SortedNames]}
        end
    catch
        throw:cycle -> 
            {error, error_message:circular_dependency()}
    after
        digraph:delete(G)
    end.

-spec add_tasks(Tasks :: [map()], G :: digraph:graph()) -> ok.
add_tasks([], _G) ->
    ok;
add_tasks([Task | Rest], G) ->
    Name = maps:get(<<"name">>, Task),
    digraph:add_vertex(G, Name, Task),
    Deps = maps:get(<<"requires">>, Task, []),
    add_edges(Name, Deps, G),
    add_tasks(Rest, G).

-spec add_edges(Name :: binary(), Deps :: [binary()], G :: digraph:graph()) -> ok.
add_edges(_, [], _G) ->
    ok;
add_edges(Name, [Dep | Rest], G) ->
    digraph:add_vertex(G, Dep),
    case digraph:add_edge(G, Dep, Name) of
        {error, {bad_edge, _}} -> throw(cycle);
        _ -> add_edges(Name, Rest, G)
    end.

-spec create_task_map(Tasks :: [map()]) -> #{binary() => map()}.
create_task_map(Tasks) ->
    maps:from_list([{maps:get(<<"name">>, Task), Task} || Task <- Tasks]).